### PR TITLE
Gate access to a thing with a Presentation Definition

### LINKED ISSUE #1: Gate access using Credentials and Presentation Exchange
# config
1. Accept a presentation definition as config
2. Register supported DID resolvers (for each credential will need to accept multiple DIDs types — especially ION)

# logic
1. Accept credentials against the PD in a Presentation Submission
2. Return true/false

### PR DESCRIPTION
**Pull request by decentralgabe**
Fixes #1

- Add a credential gate, configured with a presentation definition
- Add a method to accept a Verifiable Presentation (VP) against the presentation definition
- Automatically generate a Presentation Submission given a VP

Note: presentation exchange has limited support so the test is limited.

### CODE REVIEW

File: gate/gate.go [hunk 1]
```diff
@@ -1 +1,113 @@
...
+		resolver: r,
+		config:   config,
+	}, nil
+}
+
+func localResolverMethods() []didsdk.Method {
+	return []didsdk.Method{didsdk.KeyMethod, didsdk.WebMethod, didsdk.PKHMethod, didsdk.PeerMethod}
+}
+
+func (cg *CredentialGate) ValidatePresentationSubmission(presentationSubmissionJWT string) (bool, error) {
```
Comment (nitro-neal) [inline gate/gate.go:71]: Is the 'golang' way to only return (error) and if its nil then its valid?

File: gate/gate_test.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+			IssuanceDate: time.Now().Format(time.RFC3339),
+			CredentialSubject: map[string]any{
+				"id": didKey.String(),
+			},
+		}
+		testVCJWT, err := signing.SignVerifiableCredentialJWT(*signer, testCredential)
+		assert.NoError(tt, err)
+		assert.NotEmpty(tt, testVCJWT)
+
+		_, _, vsJSON, err := signing.ParseVerifiableCredentialFromJWT(string(testVCJWT))
```
Comment (nitro-neal) [inline gate/gate_test.go:221]: vcJSON?

File: resolver/resolver.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	return &Resolver{
+		lr: lr,
+		ur: ur,
+	}, nil
+}
+
+// Resolve resolves a DID using a combination of local and universal resolvers. The ordering is as follows:
+// 1. Try to resolve with the local resolver
+// 2. Try to resolve with the universal resolver
+func (r *Resolver) Resolve(ctx context.Context, did string, opts ...didsdk.ResolutionOption) (*didsdk.ResolutionResult, error) {
```
Comment (nitro-neal) [inline resolver/resolver.go:51]: can't an ION did be resolved locally (long form) and remotely ? 
The remote would be the source of truth?

File: resolver/universal.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+		return nil, errors.Wrap(err, "unmarshalling JSON")
+	}
+	return &result, nil
+}
+
+// Methods returns the methods that this resolver supports
+// as per https://github.com/decentralized-identity/universal-resolver/blob/main/swagger/api.yml#L121
+func (ur *universalResolver) Methods() []didsdk.Method {
+	// check if we've cached the methods
+	if len(ur.supportedMethods) > 0 {
```
Comment (nitro-neal) [inline resolver/universal.go:95]: how do we update the cache. Will this never update until someone does a new SSI-SDK()?
Maybe add a TTL timestamp to invalidate after so long

**Review by nitro-neal — APPROVED** [review]
Very good!
If this becomes a super common use case it can move into a special section of SSI-SDK

File: gate/gate.go (hunk 1 above)
Comment (decentralgabe) [reply gate/gate.go:71]: yes

File: gate/gate_test.go (hunk 2 above)
Comment (decentralgabe) [reply gate/gate_test.go:221]: renamed

File: resolver/resolver.go (hunk 3 above)
Comment (decentralgabe) [reply resolver/resolver.go:51]: you're correct, but the sdk doesn't have a local ion resolver so we'll just rely on the universal resolver. this is something good to add...
opened an issue https://github.com/TBD54566975/ssi-sdk/issues/349

File: resolver/universal.go (hunk 4 above)
Comment (decentralgabe) [reply resolver/universal.go:95]: good idea

Comment (decentralgabe) [reply resolver/universal.go:95]: this looks promising, I'll create an issue for follow up https://github.com/jellydator/ttlcache

Comment (decentralgabe) [reply resolver/universal.go]: #8

**Pull request merged**